runtime/memprof.c

```diff
@@ -156,1176 +1316,998 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
-    struct tracked* t = &entries_global.t[callback_idx];
+/* Run any pending callbacks for the current thread and domain, and
+ * any orphaned callbacks.
+ *
+ * Does not use domain_apply_actions() because this can dynamically
+ * change the various indexes into an entries table while iterating
+ * over it, whereas domain_apply_actions assumes that can't happen. */
 
-    if (t->deleted || t->running != NULL) {
-      /* This entry is not ready. Ignore it. */
-      callback_idx++;
-    } else if (t->promoted && !t->cb_promote_called) {
-      t->cb_promote_called = 1;
-      res = run_callback_exn(&entries_global, callback_idx, Promote(tracker),
-                             t->user_data);
-      if (Is_exception_result(res)) goto end;
-    } else if (t->deallocated && !t->cb_dealloc_called) {
-      value cb = (t->promoted || !t->alloc_young) ?
-        Dealloc_major(tracker) : Dealloc_minor(tracker);
-      t->cb_dealloc_called = 1;
-      res = run_callback_exn(&entries_global, callback_idx, cb, t->user_data);
+value caml_memprof_run_callbacks_exn(void)
+{
+  memprof_domain_t domain = Caml_state->memprof;
+  CAMLassert(domain);
+  memprof_thread_t thread = domain->current;
+  CAMLassert(thread);
+  value res = Val_unit;
+  if (thread->suspended || !domain->pending) return res;
+
+  orphans_adopt(domain);
+  update_suspended(domain, true);
+
+  /* run per-domain callbacks first */
+  res = entries_run_callbacks_exn(thread, &domain->entries);
+  if (Is_exception_result(res)) goto end;
+
+  /* run per-thread callbacks for current thread */
+  res = entries_run_callbacks_exn(thread, &thread->entries);
+  if (Is_exception_result(res)) goto end;
+  /* Move any surviving entries from allocating thread to owning
+   * domain, so their subsequent callbacks may be run by any thread in
+   * the domain. entries_run_callbacks_exn didn't return an exception,
+   * so all these entries have had their allocation callbacks run. If
+   * this fails due to allocation failure, the entries remain with the
+   * thread, which is OK. */
+  (void)entries_transfer(&thread->entries, &domain->entries);
+
+  /* clear out discarded/empty tables from orphans list before
+   * running callbacks. */
+  orphans_update_pending(domain);
+
+  /* now run per-domain orphaned callbacks. */
+  memprof_orphan_table_t ot = domain->orphans;
+  while (ot) {
+    entries_t es = &ot->entries;
+    if (validated_config(es) == CONFIG_NONE) {
+      /* a discarded entry table which survived orphans_update_pending
+       * must have only active entries. So we skip it. */
+      ot = ot->next;
+    } else if (es->active < es->size) {
+      res = entries_run_callbacks_exn(thread, es);
       if (Is_exception_result(res)) goto end;
+      /* orphan tables may be deallocated during callbacks (if a
+       * callback discards the profile and then orphans_update_pending
+       * runs due to a GC) so we must now start again from the
+       * begining of the list. */
```

Comment:
I don't think this is possible since there will be a running entry in the table (the one we are running), and `entries_clear_inactive` is careful not deleting running entries.

Reply:
You are quite right. The changes to `orphans_update_pending` have paid off here.